### PR TITLE
feat(pipelined): Pipelined side for polling shards

### PR DESCRIPTION
**Pull request by veshkemburu**
Signed-off-by: veshkemburu <veshkemburu@fb.com>

## Summary

For our new sharding model, pipelined side needs to be updated to be able to process shard ids. We need to add these to ActivateFlowsRequest and update the snapshots for pipelined tests. 

## Test Plan

Run pipelined unit tests.

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by ulaskozat** [comment]
what is the performance test plan for this sharding feature? We need a before and after type comparison at different load levels.

File: lte/gateway/python/magma/pipelined/tests/snapshots/test_enforcement_stats.EnforcementStatsTest.test_rule_deactivation.snapshot [hunk 1]
```diff
@@ -1,5 +1,6 @@
  cookie=0x0, table=mme(main_table), n_packets=4096, n_bytes=139264, priority=65535,ip,nw_src=192.168.128.74 actions=load:0x7594587a06d->OXM_OF_METADATA[],load:0x1->NXM_NX_REG1[],resubmit(,enforcement(main_table))
  cookie=0x0, table=mme(main_table), n_packets=0, n_bytes=0, priority=65535,ip,nw_dst=192.168.128.74 actions=load:0x7594587a06d->OXM_OF_METADATA[],load:0x10->NXM_NX_REG1[],resubmit(,enforcement(main_table))
+ cookie=0x0, table=enforcement(main_table), n_packets=128, n_bytes=4352, priority=65532,ip,reg1=0x1,metadata=0x7594587a06d,nw_src=192.168.128.74,nw_dst=45.10.0.0/25 actions=note:b'rule1',set_field:0x1->reg2,set_field:0x1->reg4,resubmit(,enforcement_stats(main_table)),resubmit(,egress(main_table))
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/tests/snapshots/test_enforcement_stats.EnforcementStatsTest.test_rule_deactivation.snapshot]: hmm I don't like this, can you rerun this test and check that its not a flake

File: lte/gateway/python/scripts/pipelined_cli.py [hunk 2]
```diff
@@ -558,6 +560,9 @@ def create_enforcement_parser(apps):
     subcmd.add_argument('--cookie', type=int, default=0)
     subcmd.add_argument('--cookie_mask', type=int, default=0)
     subcmd.set_defaults(func=get_stats_rpc)
+    subcmd = subparsers.add_parser('activate_flows',help='Activate Flows')
```
Comment (koolzz) [inline lte/gateway/python/scripts/pipelined_cli.py]: rm this line

File: lte/gateway/python/scripts/pipelined_cli.py [hunk 3]
```diff
@@ -558,6 +560,9 @@ def create_enforcement_parser(apps):
...
+    subcmd.add_argument('--shard_id', type=int, default=0, help = 'desired cookie to query for')
+    subcmd.set_defaults(func=activate_flows)
```
Comment (koolzz) [inline lte/gateway/python/scripts/pipelined_cli.py]: rm this line

File: lte/gateway/python/scripts/pipelined_cli.py [hunk 4]
```diff
@@ -558,6 +560,9 @@ def create_enforcement_parser(apps):
     subcmd.add_argument('--cookie', type=int, default=0)
     subcmd.add_argument('--cookie_mask', type=int, default=0)
     subcmd.set_defaults(func=get_stats_rpc)
+    subcmd = subparsers.add_parser('activate_flows',help='Activate Flows')
+    subcmd.add_argument('--shard_id', type=int, default=0, help = 'desired cookie to query for')
```
Comment (koolzz) [inline lte/gateway/python/scripts/pipelined_cli.py]: move this line to ~475

Comment (koolzz) [reply lte/gateway/python/scripts/pipelined_cli.py]: (we already define an activate flows parser)

File: lte/protos/pipelined.proto [hunk 5]
```diff
@@ -92,6 +92,7 @@ message ActivateFlowsRequest {
   uint32 uplink_tunnel = 9; // Subscriber uplink tunnel
   uint32 downlink_tunnel = 10; // Subscriber downlink tunnel
   repeated VersionedPolicy policies = 11; // Policy + version pairs
+  uint32 shard_id = 12; // shard id for UE
```
Comment (koolzz) [inline lte/protos/pipelined.proto]: [nit] Shard ID for UE

File: lte/gateway/python/magma/pipelined/tests/test_pull_stats.py [hunk 6]
```diff
@@ -181,11 +181,7 @@ def test_poll(self):
             self.assertEqual(rule_record_cookie.bytes_tx, 0)
             self.assertEqual(rule_record_cookie.bytes_rx, 0)
             rule_map_cookie_and_mask = self.enforcement_stats_controller.get_stats(1, 1)
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/tests/test_pull_stats.py]: Just make this get_stats(0,0), then you can keep the lovely asserts

File: lte/gateway/python/magma/pipelined/app/policy_mixin.py [hunk 7]
```diff
@@ -164,7 +164,7 @@ def _get_classify_rule_flow_msgs(self, imsi, msisdn: bytes, uplink_tunnel: int,
                     passthrough_actions,
                     hard_timeout=hard_timeout,
                     priority=Utils.UNCLASSIFIED_ALLOW_PRIORITY,
-                    cookie=rule_num,
+                    cookie=shard_id,
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/policy_mixin.py]: I think lets not modify cookie here, we don't care about other tables for polling shards so lets keep
cookie=rule_num in this file

File: lte/gateway/python/magma/pipelined/tests/test_pull_stats.py (hunk 6 above)
Comment (veshkemburu) [reply lte/gateway/python/magma/pipelined/tests/test_pull_stats.py]: get_stats(0, 0) is already tested(default arguments test), so I'll just remove this case.

File: lte/protos/pipelined.proto [hunk 8]
```diff
@@ -92,6 +92,7 @@ message ActivateFlowsRequest {
   uint32 uplink_tunnel = 9; // Subscriber uplink tunnel
   uint32 downlink_tunnel = 10; // Subscriber downlink tunnel
   repeated VersionedPolicy policies = 11; // Policy + version pairs
+  uint32 shard_id_for_ue = 12;
```
Comment (koolzz) [inline lte/protos/pipelined.proto]: mm I just meant the [nit] for the comment sorry if it wasn't clear
```
uint32 shard_id = 12; // Shard ID for UE
```

**Review by koolzz — APPROVED** [review]
Thanks for the small fixes, lgtm

**Review by pshelar — APPROVED** [review]
LGTM

**Pull request merged**